### PR TITLE
Implement Tunix-based DPO/ORPO integration.

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

Tunix-based DPO/ORPO implementation.

This DPO implementation is based on `train_sft.py`. The hooks are shared with SFT (see #3862).

Only HF datasets are currently supported (no grain and TFDS) yet.

After this PR I plan to follow up with:
* Additional documentation in the form of Jupyter notebooks
* End-to-end tests.
* Delete of the legacy DPO implementation.
* Add support of grain and TFDS datasets.

FIXES: b/485626968

# Tests

The DPO and ORPO integration was verified on a single-node TPU v5-8 VM using `Qwen2.5-1.5B-Instruct` as the base model and the `argilla/distilabel-intel-orca-dpo-pairs` dataset. 

* **E2E Functional Correctness:** Verified that the training loop runs to completion without error using the provided `train_dpo.py` script.
* **Instruction-Following Improvements (IFEval):** Evaluation over the IFEval dataset shows measurable alignment improvements:
  * **Strict Instruction Accuracy:** Improved by **+1.44%** (53.24% DPO vs. 51.80% baseline).
  * **Strict Prompt Accuracy:** Improved by **+0.37%** (41.77% DPO vs. 41.40% baseline).
* **General Knowledge Stability (MMLU):** Evaluation over the full 3,511 sharded MMLU test set confirms that general reasoning and knowledge capacity are mostly preserved post-alignment:
  * **Baseline:** 60.51%
  * **DPO Aligned:** 60.38% (99.7% knowledge retention; negligible -0.13% delta)
 
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/trainers/post_train/sft/hooks.py [hunk 1]
```diff
@@ -170,6 +167,25 @@ def on_eval_step_end(self, train_ctx: peft_trainer.PeftTrainer, eval_loss: float
...
 
+  def get_total_weights(self, batch) -> jax.Array:
+    """Calculate the number of non-padded tokens in the batch.
+
+    DPOTrainingHooks overrides this method.
+    """
+    return jnp.sum(batch["targets_segmentation"] != 0)
+
+
+class DPOTrainingHooks(SFTTrainingHooks):
```
Comment (SurbhiJainUSC) [inline src/maxtext/trainers/post_train/sft/hooks.py]: Please move this class to `src/maxtext/trainers/post_train/dpo/hooks.py`

File: src/maxtext/trainers/post_train/sft/hooks.py [hunk 2]
```diff
@@ -170,6 +167,25 @@ def on_eval_step_end(self, train_ctx: peft_trainer.PeftTrainer, eval_loss: float
...
+  """Training hooks for DPO.
```
Comment (SurbhiJainUSC) [inline src/maxtext/trainers/post_train/sft/hooks.py]: Also, we can move common functionalities from `SFTTrainingHooks` and `SFTDataHooks` to `maxtext/trainers/post_train/hooks.py` and create child classes:
1. `SFTTrainingHooks` and `SFTDataHooks` in `post_train/sft` that inherits
2. `DPOTrainingHooks` and `DPODataHooks` in `post_train/dpo` that inherits
This way, we could reuse those hooks implementation for RL hooks as well in future.

File: src/maxtext/configs/post_train/dpo.yml [hunk 3]
```diff
@@ -1,8 +1,13 @@
 base_config: "base.yml"
 
+# To use ORPO, flip these flags.
 use_dpo: true
-train_data_columns: ['chosen', 'rejected']
-eval_data_columns: ['chosen', 'rejected']
+use_orpo: false
```
Comment (SurbhiJainUSC) [inline src/maxtext/configs/post_train/dpo.yml:5]: Instead of `use_orpo`, can we have a nested config like https://github.com/AI-Hypercomputer/maxtext/blob/84bf0dfc4dfe7eb9436155afdff3d28600947c02/src/maxtext/configs/post_train/rl.yml#L37
```
dpo:
   algo: dpo or orpo
   dpo_beta
   orpo_lambda
    ....
```

File: src/maxtext/input_pipeline/input_pipeline_utils.py
Comment (SurbhiJainUSC) [inline src/maxtext/input_pipeline/input_pipeline_utils.py:1]: This file is growing. Let's keep DPO specific changes in a new module `input_pipeline/dpo_utils.py`

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 4]
```diff
@@ -623,6 +623,44 @@ def map(self, element):
...
+    if left:
+      pad_width = ((pad_amount, 0),)
+    else:
+      pad_width = ((0, pad_amount),)
+    return np.pad(x[:length], pad_width, constant_values=self.pad_id)
+
+  def map(self, element):
+    """Maps original DPO columns to Tunix-compatible pre-tokenized format."""
+    # Handle the 'input' -> 'prompt_ids' mapping
+    prompt_ids = self._pad(element.pop("input"), self.max_prompt_length, left=True)
```
Comment (SurbhiJainUSC) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: Please raise `KeyError` exception when the required key is not found in `element`

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 5]
```diff
@@ -623,6 +623,44 @@ def map(self, element):
...
+    chosen_ids = self._pad(element.pop("chosen"), self.max_response_length, left=False)
+    rejected_ids = self._pad(element.pop("rejected"), self.max_response_length, left=False)
```
Comment (SurbhiJainUSC) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: Can we not hardcode the keys like `chosen`, `rejected`?

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 6]
```diff
@@ -623,6 +623,44 @@ def map(self, element):
     return element
 
 
+class DPOTunixPrep(grain.MapTransform):
```
Comment (SurbhiJainUSC) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: Use `@dataclasses.dataclass` to simplify the initialization

File: docs/tutorials/posttraining/dpo.md
Comment (SurbhiJainUSC) [inline docs/tutorials/posttraining/dpo.md:1]: Can we also add a jupyter notebook for DPO/ORPO, which can run on Github CI to validate the implementation.

File: src/maxtext/trainers/post_train/sft/hooks.py (hunk 1 above)
Comment (igorts-git) [reply src/maxtext/trainers/post_train/sft/hooks.py]: done

File: src/maxtext/trainers/post_train/sft/hooks.py (hunk 2 above)
Comment (igorts-git) [reply src/maxtext/trainers/post_train/sft/hooks.py]: done

File: src/maxtext/input_pipeline/input_pipeline_utils.py
Comment (igorts-git) [reply src/maxtext/input_pipeline/input_pipeline_utils.py:1]: done

File: src/maxtext/input_pipeline/input_pipeline_utils.py (hunk 4 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/input_pipeline_utils.py]: done

File: src/maxtext/input_pipeline/input_pipeline_utils.py (hunk 5 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/input_pipeline_utils.py]: Yes, in fact I had a follow up PR for that and also to allow for datasets where there are only 2 columns: "chosen" and "rejected" where the "prompt" is inferred automatically by looking at the common prefix in "chosen" and "rejected". I cherry-picked that PR into this one.

File: src/maxtext/input_pipeline/input_pipeline_utils.py (hunk 6 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/input_pipeline_utils.py]: done (moved to a new file now)

File: docs/tutorials/posttraining/dpo.md
Comment (igorts-git) [reply docs/tutorials/posttraining/dpo.md:1]: I just added the notebook (vibe-coded), but I don't want to add it to Github CI yet. I still have a few follow up PRs to clean things up.

File: src/maxtext/configs/post_train/dpo.yml (hunk 3 above)
Comment (igorts-git) [reply src/maxtext/configs/post_train/dpo.yml:5]: This is a great idea, but I would rather postpone this until a later PR. First I want to clean up all the legacy DPO code, which would come as a follow up PR.

File: docs/tutorials/posttraining/dpo.md
Comment (igorts-git) [reply docs/tutorials/posttraining/dpo.md:1]: Due to the size of this PR the notebooks will be added in https://github.com/AI-Hypercomputer/maxtext/pull/3893

File: src/maxtext/configs/post_train/dpo.yml (hunk 3 above)
Comment (igorts-git) [reply src/maxtext/configs/post_train/dpo.yml]: done as part of previous PR (already merged)

File: src/maxtext/input_pipeline/input_pipeline_utils.py (hunk 5 above)
Comment (igorts-git) [reply src/maxtext/input_pipeline/input_pipeline_utils.py]: done as part of previous PR that is already merged.

**Review by SurbhiJainUSC — APPROVED** [review]

**Review by darisoy — APPROVED** [review]
